### PR TITLE
Bluetooth: controller: Data Length Update port to ULL/LLL arch

### PR DESCRIPTION
**Pull request by cvinayak**
Ported implementation of Data Length Update Procedure to
ULL/LLL architecture of the Link Layer.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 1]
```diff
@@ -395,6 +372,61 @@ u8_t ll_version_ind_send(u16_t handle)
...
+	return 0;
+}
+
+void ll_length_max_get(u16_t *max_tx_octets, u16_t *max_tx_time,
+		       u16_t *max_rx_octets, u16_t *max_rx_time)
+{
+	*max_tx_octets = LL_LENGTH_OCTETS_RX_MAX;
+	*max_tx_time = PKT_US(LL_LENGTH_OCTETS_RX_MAX, BIT(2));
+	*max_rx_octets = LL_LENGTH_OCTETS_RX_MAX;
+	*max_rx_time = PKT_US(LL_LENGTH_OCTETS_RX_MAX, BIT(2));
```
Comment (mped-oticon) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]: BIT(2) means coded, which of course is a safe, but not very tight, bound. 
So this should be `#ifdef`'ed on whether coded phy is supported.

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 2]
```diff
@@ -2458,12 +2519,12 @@ static inline void event_len_prep(struct ll_conn *conn)
 
 		lr = &pdu_ctrl_tx->llctrl.length_req;
 		lr->max_rx_octets = LL_LENGTH_OCTETS_RX_MAX;
-		lr->max_tx_octets = lll->default_tx_octets;
+		lr->max_tx_octets = conn->default_tx_octets;
 		lr->max_rx_time = PKT_US(LL_LENGTH_OCTETS_RX_MAX, BIT(2));
```
Comment (mped-oticon) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]: This always assumes coded phy.

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 3]
```diff
@@ -3567,6 +3574,240 @@ static void reject_ext_ind_recv(struct ll_conn *conn, struct node_rx_pdu *rx,
...
+#endif /* CONFIG_BT_CTLR_PHY_CODED */
+		}
+
+		/* use the minimal of our max supported and
+		 * peer max_tx_time
+		 */
+		if (lr->max_tx_time >= PKT_US(PDU_DC_PAYLOAD_SIZE_MIN, 0)) {
+			eff_rx_time = MIN(lr->max_tx_time,
+					  PKT_US(LL_LENGTH_OCTETS_RX_MAX,
+						 BIT(2)));
```
Comment (mped-oticon) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]: doesn't BIT(2) imply coded phy? Yet it is outside `#if defined(CONFIG_BT_CTLR_PHY_CODED)`

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 4]
```diff
@@ -1202,6 +1255,22 @@ void ull_conn_tx_ack(struct ll_conn *conn, memq_link_t *link,
 	ll_tx_ack_put(conn->lll.handle, tx);
 }
 
+u8_t ull_conn_llcp_req(void *conn)
+{
+	struct ll_conn * const conn_hdr = conn;
+	if (conn_hdr->llcp_req != conn_hdr->llcp_ack) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
```
Comment (mped-oticon) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]: This makes good sense. req != ack, so we have an ongoing/pending operation, thus we shall deny this new request.

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 5]
```diff
@@ -1202,6 +1255,22 @@ void ull_conn_tx_ack(struct ll_conn *conn, memq_link_t *link,
...
+	}
+
+	conn_hdr->llcp_req++;
```
Comment (mped-oticon) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]: Here req == ack, so we shall grant the new request. Fine.

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 6]
```diff
@@ -1202,6 +1255,22 @@ void ull_conn_tx_ack(struct ll_conn *conn, memq_link_t *link,
...
+	if (((conn_hdr->llcp_req - conn_hdr->llcp_ack) & 0x03) != 1) {
```
Comment (mped-oticon) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]: Given above control flow, req will always be one higher than ack (modulo 256 (which here is the same as modulo 4)), so this condition will always be false.

File: subsys/bluetooth/controller/ll_sw/ull_conn.c [hunk 7]
```diff
@@ -1202,6 +1255,22 @@ void ull_conn_tx_ack(struct ll_conn *conn, memq_link_t *link,
...
+		conn_hdr->llcp_req--;
+		return BT_HCI_ERR_CMD_DISALLOWED;
```
Comment (mped-oticon) [inline subsys/bluetooth/controller/ll_sw/ull_conn.c]: This then-body is dead code?

**Review by mped-oticon — COMMENTED** [review]
Looks good, but have a few comments.
Also not completed full review; will await response before continuing.

File: subsys/bluetooth/controller/ll_sw/ull_conn.c (hunk 7 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_conn.c]: as ISR can request a control procedure, which then is doing llcp_ack-- between the lines 1265 and 1266, the then body is required to detect the race-condition.

**Comment by cvinayak** [comment]
@mped-oticon updated, please re-review.

**Review by mped-oticon — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c [hunk 8]
```diff
@@ -614,6 +614,12 @@ void lll_conn_pdu_tx_prep(struct lll_conn *lll, struct pdu_data **pdu_data_tx)
 		}
 	}
 
+	p->rfu = 0U;
+
+#if !defined(CONFIG_BT_CTLR_DATA_LENGTH_CLEAR)
+	p->resv = 0U;
```
Comment (cristic-devel) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c]: What is the purspose of this byte in the PDU data structure?

**Pull request merged**